### PR TITLE
Support parsing for attribute and tuple level constraint

### PR DESCRIPTION
**Pull request by yliuuuu**
## Relevant Issues
- [Closes/Related To] Issue #XXX

## Description
- Supports parsing of Attribute-Level Constraint (CHECK, NULL, NOT NULL, UNIQUE, PRIMARY KEY)
- Supports parsing of Table-Level Constraints (CHECK, UNIQUE, PRIMARY KEY)
- It contains breaking changes on the AST modeling. in particular remodeling of constraints to be unnested from table definition. 
- This is because the constraint can be reused in multiple parts of the DDL operation, (in this PR, attribute level constraint and table level constraints; in the future, we might needs to use constraint in ALTER Statement). 

### Added
- Supports parsing of Attribute-Level Constraint (CHECK, NULL, NOT NULL, UNIQUE, PRIMARY KEY)
- Supports parsing of Table-Level Constraints (CHECK, UNIQUE, PRIMARY KEY)
### Changed
- Modeling of DDL node. 
    - DDL node is modeled as a wrapper around DDLOp. For example: replace `statementDDLCreateTable(...)` to `statementDDL(ddlOpCreateTable(....))`

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES/NO]**
  -No. Targeting V1 branch. 

- Any backward-incompatible changes? **[YES/NO]**
  -  Yes. AST modeling. 
  
- Any new external dependencies? **[YES/NO]**
  - No. 
  
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **[YES/NO]**
- Yes. 

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-ast/src/main/resources/partiql_ast.ion [hunk 1]
```diff
@@ -785,23 +788,27 @@ returning::{
...
-          ],
-        },
-      ],
+    }
+  ],
+}
+
+constraint::{
+  name: optional::string,
+  body: [
```
Comment (alancai98) [inline partiql-ast/src/main/resources/partiql_ast.ion:804]: (naming) not sure if `body` is a good name. Perhaps could use `type` or something else?

File: partiql-parser/src/test/kotlin/org/partiql/parser/internal/PartiQLParserDDLTests.kt [hunk 2]
```diff
@@ -77,27 +107,232 @@ class PartiQLParserDDLTests {
                     null
                 )
             ),
+
+            //
```
Comment (alancai98) [inline partiql-parser/src/test/kotlin/org/partiql/parser/internal/PartiQLParserDDLTests.kt:111]: nice set of tests. could also consider adding a few more:
- multiple constraints
- search condition w/ an `OR` (like what was mentioned in the ANTLR grammar
- other parsing error edge cases
  - missing parens
  - table constraint used where attribute constraint should be placed (and vice-versa)

File: partiql-lang/src/test/kotlin/org/partiql/lang/syntax/PartiQLParserDDLTest.kt [hunk 3]
```diff
@@ -35,7 +35,47 @@ internal class PartiQLParserDDLTest : PartiQLParserTestBase() {
                 query = "DROP Table foo.bar",
                 code = ErrorCode.PARSE_UNEXPECTED_TOKEN,
                 context = mapOf(),
-            )
+            ),
```
[the remark this answers is not in the document]
Comment (yliuuuu) [reply partiql-lang/src/test/kotlin/org/partiql/lang/syntax/PartiQLParserDDLTest.kt:38]: Probably not, as the old parser is getting deprecated. 

Those tests are set to ignore context, hence the context here is an empty map. But in Pig visitor code an error message is included.

File: partiql-parser/src/test/kotlin/org/partiql/parser/internal/PartiQLParserDDLTests.kt (hunk 2 above)
Comment (yliuuuu) [reply partiql-parser/src/test/kotlin/org/partiql/parser/internal/PartiQLParserDDLTests.kt:111]: Good suggestion, we can add more tests once the parser portion of DDL implementation is completed. Since we are going to add more features, the modeling will change hence the test code.

File: partiql-ast/src/main/resources/partiql_ast.ion (hunk 1 above)
Comment (yliuuuu) [reply partiql-ast/src/main/resources/partiql_ast.ion:804]: I do not like the name either. 

To address this and other naming suggestion in the comments: 

The name are reused from the previous implementation. I am more than happy to change those: 

"column" -> "attribute" 
"body" -> "definition"

File: partiql-ast/src/main/resources/partiql_ast.ion [hunk 4]
```diff
@@ -785,23 +788,27 @@ returning::{
...
-            check::{ expr: expr },
-          ],
-        },
-      ],
+    }
+  ],
+}
+
+constraint::{
+  name: optional::string,
```
[the remark this answers is not in the document]
Comment (yliuuuu) [reply partiql-ast/src/main/resources/partiql_ast.ion:803]: This is a good call out. 

Sql's grammar is constructed this way because they treat constraint as an object... 

The semantics with constraint name is not clear at the moment, but constraint name was modeled as a string previously in our grammar.

File: partiql-parser/src/main/kotlin/org/partiql/parser/internal/PartiQLParserDefault.kt [hunk 5]
```diff
@@ -588,62 +592,94 @@ internal class PartiQLParserDefault : PartiQLParser {
...
 
         override fun visitCreateTable(ctx: GeneratedParser.CreateTableContext) = translate(ctx) {
             val table = visitQualifiedName(ctx.qualifiedName())
             val definition = ctx.tableDef()?.let { visitTableDef(it) }
-            statementDDLCreateTable(table, definition)
+            ddlOpCreateTable(table, definition)
         }
 
         override fun visitCreateIndex(ctx: GeneratedParser.CreateIndexContext) = translate(ctx) {
             // TODO add index name to ANTLR grammar
```
Comment (alancai98) [inline partiql-parser/src/main/kotlin/org/partiql/parser/internal/PartiQLParserDefault.kt:617]: Is this TODO still relevant based on the updated DDL modeling?

File: partiql-parser/src/main/antlr/PartiQL.g4 [hunk 6]
```diff
@@ -100,17 +99,43 @@ tableDef
...
 
 columnConstraint
-    : ( CONSTRAINT columnConstraintName )?  columnConstraintDef
+    : ( CONSTRAINT constraintName )?  columnConstraintDef
     ;
 
 columnConstraintDef
-    : NOT NULL                                  # ColConstrNotNull
-    | NULL                                      # ColConstrNull
+    : NOT NULL                                     # ColConstrNotNull
```
Comment (alancai98) [inline partiql-parser/src/main/antlr/PartiQL.g4:115]: Since the partiql-ast now uses "attribute" in the table definition, should the ANTLR definitions also use "attribute" rather than "column"? I think we should keep the parser and ast terminology consistent.

Related -- would the PIG parser also need to be updated to use "attribute"?

File: partiql-ast/src/main/resources/partiql_ast.ion [hunk 7]
```diff
@@ -785,23 +788,27 @@ returning::{
...
-            not_null::{},
-            check::{ expr: expr },
-          ],
-        },
-      ],
+    }
+  ],
+}
+
+constraint::{
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-ast/src/main/resources/partiql_ast.ion:800]: Based on offline discussion, there's a lot of overlap between the "table" and "attribute" constraints hence why they're modeled as one "constraint" node here. @yliuuuu had also mentioned that this unified modeling made the conversion to the plan a bit simpler. As we continue to work on the rest of DDL modeling in the AST and plan we can adjust the modeling if needed.

File: partiql-parser/src/main/kotlin/org/partiql/parser/internal/PartiQLParserDefault.kt (hunk 5 above)
Comment (yliuuuu) [reply partiql-parser/src/main/kotlin/org/partiql/parser/internal/PartiQLParserDefault.kt:617]: Yes. Current Index grammar is untouched.

File: partiql-parser/src/main/antlr/PartiQL.g4 (hunk 6 above)
Comment (yliuuuu) [reply partiql-parser/src/main/antlr/PartiQL.g4:115]: I think this is a valid call out. But given the fact that Antlr grammar is an internal concern, perhaps we can treat this as a backlog item? 

I don't think we want to introduce changes to the PIG parser.

Comment (yliuuuu) [reply partiql-parser/src/main/antlr/PartiQL.g4:115]: At some point we would want to update the ANTLR grammar on our [website](https://partiql.org/syntax/antlr.html). Perhaps then would be a good time to do those changes in batch?

Comment (alancai98) [reply partiql-parser/src/main/antlr/PartiQL.g4:115]: > I think this is a valid call out. But given the fact that Antlr grammar is an internal concern, perhaps we can treat this as a backlog item?

Ah, I hadn't realized the current ANTLR-generated parse code was meant to be "internal" despite the visibility modifiers indicating it was "public". Created an issue to track the internalization of the ANTLR-generated sources -- https://github.com/partiql/partiql-lang-kotlin/issues/1446.

> At some point we would want to update the ANTLR grammar on our [website](https://partiql.org/syntax/antlr.html). Perhaps then would be a good time to do those changes in batch?

Since we'll likely have other changes to the ANTLR grammar before the DDL syntax is finalized, I'm fine w/ punting on the changes. Could you create an issue to track updating the website grammar and grammar within `partiql-parser`? That should make it easier to track what'll need changed before the next release.

**Review by alancai98 — APPROVED** [review]
